Managers/MapManager.gd

```diff
@@ -51,27 +51,162 @@ func create_map(map_floors_width: Array[int] = map_width_array) -> MapBase:
-			# create a new room with the room type and give it its position
+			# create a new room with the null type and give it its position
 			var _generated_room: RoomBase = RoomBase.new()
-			_generated_room.room_event = _room_event
+			_generated_room.room_event = null
 			_generated_room.room_position = Vector2i(index_width + _padding_size, index_height)
 			# put the new room on the grid
-
 			_grid[index_height].append(_generated_room as RoomBase)
 		_grid[index_height].append_array(_padding)
 	_map.rooms = _grid
+
+	# Assign room type to all rooms
+	assign_events(_map)
+
 	return _map as MapBase
 
+## Checks if the current room event upholds all map generation rules
+func is_room_event_correct(current_room: RoomBase, map: MapBase = current_map) -> bool:
+	## Rule 0: A room cannot have no event
+	if current_room.room_event == null:
+		return false
+	
+	var current_room_event_name: String = current_room.room_event.get_event_name()
+
+	## Rule 1: No Heal room before half of the map
+	if current_room_event_name == "heal" && current_room.room_position.y < int(map.rooms.size()/2.0):
+		return false
+
+
+	## Rule 2: Cannot have 2 consecutive Heal or Shop rooms
+	# Room events that we do not want to be consecutive
+	const no_consecutive_room_event: Array[String] = ["shop", "heal"]
+	
+	# We need to check the parents if current room event is shop or heal
+	if current_room_event_name in no_consecutive_room_event:
+		# Get the parent rooms' types
+		var _parent_events: Array[String] = []
+		for delta_x: int in [-1, 0, 1]:
+			var _parent_x: int = current_room.room_position.x + delta_x
+			var _parent_y: int = current_room.room_position.y - 1
+			if _parent_y >= 0 && _parent_x >= 0 && _parent_x < map.rooms[_parent_y].size() && map.rooms[_parent_y][_parent_x] != null:
+				_parent_events.append(map.rooms[_parent_y][_parent_x].room_event.get_event_name())
+
+		# Check if current room has the same type as one of its parent room
+		if current_room_event_name in _parent_events:
+			return false
+
+
+	## Rule 3: There must be at least 2 room types among destinations of Rooms that have 2 or more Paths going out. [br]
+	## Since the events of the rooms on the right of current room ((x+i,y), i=1,2,...) have not been generated yet, [br]
+	## we don't need to take them into account. We only check neighbors on the left: (x-2,y), (x-1,y) and (x,y) [br]
+	## Note: We make sure the two leftmost and rightmost rooms of the floor have different events, [br]
+	## since they could be the only destinations of a room from the previous floor (happens when current floor's size [br]
+	## is smaller or equal than previous)
+	var _neighbors_events: Array[String] = []
+	var delta_xs: Array[int] = []
+
+	# If rightmost room of floor
+	if current_room.room_position.x == map.rooms[current_room.room_position.y].size() - 1 or map.rooms[current_room.room_position.y][current_room.room_position.x+1] == null:
+		delta_xs = [-1, 0]
+	else:
+		delta_xs = [-2, -1, 0]
```

Comment:
i'm not sure i see why this is done
I understand why we need to check -2, -1 and 0
but if a room is the rightmost of the floor, someone on the floor below could still access -2, -1 and 0.
Basically, when the map is getting smaller, if you are not already on the rightmost side of the map you can choose between -2, -1 or 0 for the next floor. If you are already on the rightmost/leftmost side of the map, you only have a single room choice

Reply:
oh yeah mb i was only thinking of the rightmost room, but the room before that can only access two rooms indeed, not three
All good :+1: